java/src/main/native/ai_onnxruntime_OrtSession.c

```diff
@@ -242,116 +318,159 @@ JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getOutputInfo
-    const char** outputNames;
-    checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator,sizeof(char*)*numOutputs,(void**)&outputNames));
-    jobject* javaInputStrings;
-    checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator,sizeof(jobject)*numInputs,(void**)&javaInputStrings));
-    jobject* javaOutputStrings;
-    checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator,sizeof(jobject)*numOutputs,(void**)&javaOutputStrings));
-
-    // Extract a C array of longs which are pointers to the input tensors.
-    // Need to convert longs to OrtValue* in case we run on non-64bit systems
-    jlong* inputTensors = (*jniEnv)->GetLongArrayElements(jniEnv,tensorArr,NULL);
-    const OrtValue** inputValues;
-    checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator,sizeof(OrtValue*)*numInputs,(void**)&inputValues));
-
-    // Extract the names and native pointers of the input values.
-    for (int i = 0; i < numInputs; i++) {
-        javaInputStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv,inputNamesArr,i);
-        inputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv,javaInputStrings[i],NULL);
-        inputValues[i] = (OrtValue*)inputTensors[i];
-    }
+JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_run(JNIEnv* jniEnv, jobject jobj, jlong apiHandle,
+                                                                  jlong sessionHandle, jlong allocatorHandle,
+                                                                  jobjectArray inputNamesArr, jlongArray tensorArr,
+                                                                  jlong numInputs, jobjectArray outputNamesArr,
+                                                                  jlong numOutputs, jlong runOptionsHandle) {
+
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtAllocator* allocator = (OrtAllocator*)allocatorHandle;
+  OrtSession* session = (OrtSession*)sessionHandle;
+  OrtRunOptions* runOptions = (OrtRunOptions*)runOptionsHandle;
+
+  jobjectArray outputArray = NULL;
+
+  // Create the buffers for the Java input & output strings, and the input pointers
+  const char** inputNames = malloc(sizeof(char*) * numInputs);
+  if (inputNames == NULL) {
+    // Nothing to cleanup, return and throw exception
+    return outputArray;
+  }
+  const char** outputNames = malloc(sizeof(char*) * numOutputs);
+  if (outputNames == NULL) {
+    goto cleanup_input_names;
+  }
+  jobject* javaInputStrings = malloc(sizeof(jobject) * numInputs);
+  if (javaInputStrings == NULL) {
+    goto cleanup_output_names;
+  }
+  jobject* javaOutputStrings = malloc(sizeof(jobject) * numOutputs);
+  if (javaOutputStrings == NULL) {
+    goto cleanup_java_input_strings;
+  }
+  const OrtValue** inputValuePtrs = malloc(sizeof(OrtValue*) * numInputs);
```

Comment:
>nputValuePtrs = malloc(sizeof(OrtValu [](http://example.com/codeflow?start=20&length=37)

I would not expect this to compile. malloc() returns void* which is not implicitly convertible to any pointer (the other way around is fine).
It is possible that some warnings in some settings may be off, but generally this is wrong.
I would write it this way OrtValue* const * inputValuePtrs = (OrtValue**) malloc(...

Reply:
Ok, I can update it in this PR or another later one. It compiled and ran fine on macOS with clang.